srcts/src/components/darkModeSwitch.ts

```diff
@@ -0,0 +1,323 @@
+import type { CSSResultGroup } from "lit";
+import { LitElement } from "lit";
+import { css, html } from "lit";
+import { property } from "lit/decorators.js";
+
+// Inspired by:
+// https://web.dev/building-a-theme-switch-component/
+// https://web.dev/patterns/theming/theme-switch/
+// https://github.com/argyleink/gui-challenges/tree/main/theme-switch
+
+export type DarkModeMessageToggle = {
+  method: "toggle";
+  value: DarkModeSwitch["mode"] | undefined;
+};
+
+export class DarkModeSwitch extends LitElement {
+  static isShinyInput = true;
+  static tagName = "bslib-input-dark-mode";
+  static shinyCustomMessageHandlers = {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    "bslib.toggle-dark-mode": ({
+      method,
+      value,
+    }: DarkModeMessageToggle): void => {
+      // Similar to DarkModeSwitch.receiveMessage(), but we directly update the
+      // Bootstrap attribute on the <html> element. Currently, all toggle switches
+      // follow this value.
+
+      if (method !== "toggle") return;
+
+      if (typeof value === "undefined" || value === null) {
+        const current = document.documentElement.dataset.bsTheme || "light";
+        value = current === "light" ? "dark" : "light";
+      }
+
+      document.documentElement.dataset.bsTheme = value;
+    },
+  };
```

Comment:
I think that this actually should be separate, and should go inside of `bslibShiny.ts`.

With the way that the code is currently structured, the dark mode state is part of the page. The switch component reflects the state and lets you change the state, but the `<html>` tag is the source of truth for the dark mode state. 

If someone is programmatically setting the dark mode state from R, it's plausible that they'd want to do it even if there's no dark mode switch on the page. However, if the app doesn't contain the dark mode switch (or any other components in `webComponents.js`) then this code won't be loaded, and they won't be able to programmatically set light/dark mode.

Reply:
This line of reasoning also implies that the dark mode/theme state should be available via `clientData`, but that could be done at some point in the future.

Reply:
> If someone is programmatically setting the dark mode state from R, it's plausible that they'd want to do it even if there's no dark mode switch on the page. However, if the app doesn't contain the dark mode switch (or any other components in webComponents.js) then this code won't be loaded, and they won't be able to programmatically set light/dark mode.

Unfortunately, moving this code to `bslibShiny.ts` doesn't solve this problem, since all components deps are added when the component is added to the page.

I think it makes sense to require the `input_dark_mode()` to be added to the app as a way for the app author to explicitly support dark mode, both to give the app user the opportunity to pick the color mode, as well as a way for us to explicitly check the Bootstrap version (which could be 3 or 4 and therefore not supporting dark mode).

Reply:
FWIW, we're probably going into a direction were if any bslib UI is used, then all component-specific assets (bundled into a single `htmlDependency()`) will come with it, which I think would "solve" this problem.

Reply:
@cpsievert true, and that would make the server side toggle work without the input, which I also think is okay. I'd rather keep as much of the implementation together in one place though, unless there a philosophically bigger reason why we would want to separate these pieces.

Reply:
I can see the reasons for keeping the code in this file... But, in the interest of having the component class encapsulate just the component's functionality, I think this message handler should be moved outside of the class.